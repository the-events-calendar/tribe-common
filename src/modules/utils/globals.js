/**
 * @todo: handle globals in a better way
 */

export const get = ( key, defaultValue ) => window[ key ] || defaultValue;
export const google = () => get( 'google' );
export const wpApi = wp.api;
export const wpApiRequest = wp.apiRequest;
export const wpComponents = wp.components;
export const wpData = wp.data;
export const wpEditor = wp.blockEditor || wp.editor;
export const wpHooks = wp.hooks;

export const wpDataSelectCoreEditor = () => wpData.select( 'core/block-editor' ) ||
	wpData.select( 'core/editor' );

export const IconButton = wpComponents.Button || wpComponents.IconButton;

// Localized Config
export const config = () => get( 'tribe_editor_config', {} );

// Common
export const common = () => config().common || {};
export const adminUrl = () => common().adminUrl || '';
export const rest = () => common().rest || {};
export const restNonce = () => rest().nonce || {};
export const dateSettings = () => common().dateSettings || {};
export const editorConstants = () => common().constants || {};
export const list = () => ( {
	countries: common().countries || {},
	us_states: common().usStates || {},
} );

// TEC
export const tec = () => config().events || {};
export const editor = () => tec().editor || {};
export const settings = () => tec().settings || {};
export const mapsAPI = () => tec().googleMap || {};
export const priceSettings = () => tec().priceSettings || {};
export const tecDateSettings = () => tec().dateSettings || {};
export const timezoneHtml = () => tec().timezoneHTML || '';
export const defaultTimes = () => tec().defaultTimes || {};
export const timezone = () => tec().timeZone || {};

// PRO
export const pro = () => config().eventsPRO || {};
export const editorDefaults = () => pro().defaults || {};

// Tickets
export const tickets = () => config().tickets || {};

// Tickets Plus
export const ticketsPlus = () => config().ticketsPlus || {};
export const iacVars = () => ticketsPlus().iacVars || {};

// Post Objects
export const postObjects = () => config().post_objects || {};
export const post = () => config().post || {};

// Blocks
export const blocks = () => config().blocks || {};
