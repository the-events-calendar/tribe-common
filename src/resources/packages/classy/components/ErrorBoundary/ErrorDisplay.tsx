import * as React from 'react';
import { _x } from '@wordpress/i18n';

type ErrorDisplayProps = {
	error: Error;
};

/**
 * A pure function component that displays an error message and stack trace.
 *
 * @since TBD
 *
 * @param {Error} error The error object to display.
 */
export default function ErrorDisplay( { error }: ErrorDisplayProps ): React.ReactNode {
	return (
		<div
			className="classy-root classy-root--error"
			style={ {
				alignItems: 'center',
				backgroundColor: '#fdd',
				border: '1px solid #d6e9c6',
				color: '#a94442',
				display: 'flex',
				flexDirection: 'column',
				gap: '20px',
				justifyContent: 'center',
				overflow: 'scroll',
				padding: '20px',
			} }
		>
			<h2>
				{ _x(
					'An error occurred in the Classy application:',
					'Message before the error call stack',
					'tribe-common'
				) }
			</h2>
			<pre
				style={ {
					backgroundColor: '#eee',
					borderRadius: '4px',
					marginTop: '20px',
					maxWidth: '100%',
					padding: '15px',
					textWrap: 'pretty',
				} }
			>
				{ error.stack }
			</pre>
		</div>
	);
}
