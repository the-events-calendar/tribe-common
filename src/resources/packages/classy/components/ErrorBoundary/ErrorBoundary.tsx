import * as React from 'react';
import { Component } from 'react';
import ErrorDisplay from './ErrorDisplay';
import { _x } from '@wordpress/i18n';

type ErrorBoundaryProps = {
	children: React.ReactNode;
	errorMessage?: string;
};

type ErrorBoundaryState = {
	error: Error | null;
};

const defaultErrorMessage = _x(
	'Classy has encountered an error:',
	'Error message displayed when an error occurs in the Classy editor. An error message will follow.',
	'tribe-common'
);

/**
 * A boundary component that catches JavaScript errors anywhere in its child component tree,
 * logs those errors, and displays a fallback UI.
 *
 * @since TBD
 */
export default class ErrorBoundary extends Component< ErrorBoundaryProps, ErrorBoundaryState > {
	/**
	 * Constructs the ErrorBoundary instance with the initial state.
	 *
	 * @since TBD
	 *
	 * @param {ErrorBoundaryProps} props The properties passed to this component.
	 */
	constructor( props: ErrorBoundaryProps ) {
		super( props );
		this.state = { error: null };
	}

	/**
	 * Returns the error message to be displayed when an error occurs.
	 *
	 * @since TBD
	 *
	 * @return {string} The error message to display.
	 */
	getErrorMessage(): string {
		const { errorMessage = defaultErrorMessage } = this.props;

		return errorMessage;
	}

	/**
	 * Updates the state with the error caught by the nearest descendant error boundary.
	 *
	 * @since TBD
	 *
	 * @param {Error} error The error object that was thrown.
	 *
	 * @return {ErrorBoundaryState} The new state of the component.
	 */
	static getDerivedStateFromError( error: Error ): ErrorBoundaryState {
		return { error };
	}

	/**
	 * Logs the error and its information to the console.
	 *
	 * @since TBD
	 *
	 * @param {any} error The error object that was thrown.
	 * @param {any} errorInfo Information about which component threw the error.
	 */
	componentDidCatch( error: any, errorInfo: any ) {
		console.error( this.getErrorMessage(), error, errorInfo );
	}

	/**
	 * Renders either the fallback UI if there is an error in the state,
	 * or the children components otherwise.
	 *
	 * @since TBD
	 *
	 * @return {React.ReactNode} The component to render.
	 */
	render(): React.ReactNode {
		if ( this.state.error !== null ) {
			return <ErrorDisplay error={ this.state.error } />;
		}

		return this.props.children;
	}
}
