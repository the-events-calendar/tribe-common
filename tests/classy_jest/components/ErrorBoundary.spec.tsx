import * as React from 'react';
import { render } from '@testing-library/react';
import { describe, expect, it, beforeAll, afterAll, jest } from '@jest/globals';
import ErrorBoundary from '../../../src/resources/packages/classy/components/ErrorBoundary/ErrorBoundary';

// Mock console.error to suppress output in tests
const originalError = console.error;

describe( 'ErrorBoundary Component', () => {
	beforeAll( () => {
		console.error = jest.fn();
	} );
	afterAll( () => {
		console.error = originalError;
	} );

	it( 'renders children when there is no error', () => {
		const { container } = render(
			<ErrorBoundary>
				<div>Test Child</div>
			</ErrorBoundary>
		);

		expect( container.firstChild ).toMatchSnapshot();
	} );

	it( 'catches an error and displays the ErrorDisplay', async () => {
		const ThrowingComponent = () => {
			const error = new Error( 'for reasons' );
			error.stack = '__ERROR_CALL_STACK__';
			throw error;
		};

		const { container } = render(
			<ErrorBoundary>
				<ThrowingComponent />
			</ErrorBoundary>
		);

		expect( container.firstChild ).toMatchSnapshot();
	} );
} );
