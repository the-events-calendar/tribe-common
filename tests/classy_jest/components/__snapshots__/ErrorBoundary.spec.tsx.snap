// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary Component catches an error and displays the ErrorDisplay 1`] = `
<div
  class="classy-root classy-root--error"
  style="align-items: center; background-color: rgb(255, 221, 221); border: 1px solid #d6e9c6; color: rgb(169, 68, 66); display: flex; flex-direction: column; gap: 20px; justify-content: center; overflow: scroll; padding: 20px;"
>
  <h2>
    An error occurred in the Classy application:
  </h2>
  <pre
    style="background-color: rgb(238, 238, 238); border-radius: 4px; margin-top: 20px; max-width: 100%; padding: 15px; text-wrap: pretty;"
  >
    __ERROR_CALL_STACK__
  </pre>
</div>
`;

exports[`ErrorBoundary Component renders children when there is no error 1`] = `
<div>
  Test Child
</div>
`;
